R/ci_validate_hub_config.R

```diff
@@ -0,0 +1,118 @@
+#' directory. If running interactively, it will need to be cleaned up.
+#' @details
+#' This function is to be used within a continuous integration context.
+#' It is intended to be used in a workflow that checks the validity of
+#' a hub context. Below is an example setup of steps on GitHub Actions
+#' where the environment variables `PR_NUMBER` and `HUB_PATH` have been
+#' defined:
+#'
+#' ```yaml
+#'      - uses: actions/checkout@v4
+#'      - uses: r-lib/actions/setup-r@v2
+#'        with:
+#'          install-r: false
+#'          use-public-rspm: true
+#'          extra-repositories: 'https://hubverse-org.r-universe.dev'
+#'      - uses: r-lib/actions/setup-r-dependencies@v2
+#'        with:
+#'          cache: 'always'
+#'          packages: |
+#'            any::hubAdmin
+#'            any::sessioninfo
+#'      - name: Run validations
+#'        id: validate
+#'        run: hubAdmin::ci_validate_config()
+#'        shell: Rscript {0}
+#'      - name: "Comment on PR"
+#'        id: comment-diff
+#'        if: ${{ github.event_name != 'workflow_dispatch' }}
+#'        uses: carpentries/actions/comment-diff@main
+#'        with:
+#'          pr: ${{ env.PR_NUMBER }}
+#'          path: ${{ env.HUB_PATH }}/diff.md
+#'      - name: Error on Failure
+#'        if: ${{ steps.validate.outputs.result == 'false' }}
+#'        run: |
+#'          echo "::error title=Invalid Configuration::Errors were detected"
+#'          exit 1
+#' ```
+#'
+#' @note This function is not intended for interactive use.
+#'
+#' @keywords internal
+#' @export
+#' @examples
+#' #
+#' tmp <- tempfile()
+#' out <- tempfile()
+#' dir.create(tmp)
+#' # Results from a valid hub -----------------------------------------
+#' file.copy(
+#'   from = system.file("testhubs/simple/", package = "hubUtils"),
+#'   to = tmp,
+#'   recursive = TRUE
+#' )
+#' hub <- file.path(tmp, "simple")
+#' ci_validate_hub_config(hub_path = hub, gh_output = out)
+#' # result is true
+#' writeLines(readLines(out))
+#' # message to user shows success and a timestamp
+#' writeLines(readLines(file.path(hub, "diff.md")))
+#'
+#' # Results from an invalid hub --------------------------------------
+#' # reset output file
+#' out <- tempfile()
+#' # make the the simple hub invalid by adding a character where
+#' # a number should be
+#' tasks_path <- file.path(hub, "hub-config", "tasks.json")
+#' tasks <- readLines(tasks_path)
+#' writeLines(sub('minimum": 0', 'minimum": "0"', tasks), tasks_path)
+#' # validate
+#' ci_validate_hub_config(hub_path = hub, gh_output = out)
+#' # result is now false
+#' writeLines(readLines(out))
+#' # message to user now shows a table
+#' writeLines(head(readLines(file.path(hub, "diff.md"))))
+#' writeLines(tail(readLines(file.path(hub, "diff.md"))))
+ci_validate_hub_config <- function(hub_path = Sys.getenv("HUB_PATH"), gh_output = Sys.getenv("GITHUB_OUTPUT"), ...) {
+  diff <- file.path(hub_path, "diff.md")
```

Comment:
I know this doesn't really affect anything when used in CI and you do mention that it's not designed to be used interactively, but given it does write out in interactive use and there's nothing really stopping anyone from doing so, is there a reason we write this out within the `hub_path`? 

I believe CRAN doesn't love functions that write stuff out too without being very clear and explicit about that being the purpose of the function. Given that writing out a `diff.md` is not the explicit purpose of the function, can we not use a temp file instead? Or atleast add an `on.exit()` expression that cleans the `diff.md` file up automatically?

Reply:
done